subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/ProviderConnection.java

```diff
@@ -137,22 +143,48 @@ public Object runClientAction(Object clientAction, BuildCancellationToken cancel
         StartParameterInternal startParameter = new ProviderStartParameterConverter().toStartParameter(providerParameters, params.properties);
         ProgressListenerConfiguration listenerConfig = ProgressListenerConfiguration.from(providerParameters);
         BuildAction action = new ClientProvidedBuildAction(startParameter, serializedAction, tasks != null, listenerConfig.clientSubscriptions);
-        return run(action, cancellationToken, listenerConfig, providerParameters, params);
+        return run(action, cancellationToken, listenerConfig, listenerConfig.buildEventConsumer, providerParameters, params);
+    }
 
+    public Object runPhasedAction(InternalPhasedAction clientPhasedAction,
+                                  PhasedActionResultListener resultListener,
+                                  BuildCancellationToken cancellationToken,
+                                  ProviderOperationParameters providerParameters) {
+        SerializedPayload serializedAction = payloadSerializer.serialize(clientPhasedAction);
+        Parameters params = initParams(providerParameters);
+        StartParameterInternal startParameter = new ProviderStartParameterConverter().toStartParameter(providerParameters, params.properties);
+        FailsafePhasedActionResultListener failsafePhasedActionResultListener = new FailsafePhasedActionResultListener(resultListener);
+        ProgressListenerConfiguration listenerConfig = ProgressListenerConfiguration.from(providerParameters);
+        BuildAction action = new ClientProvidedPhasedAction(startParameter, serializedAction, listenerConfig.clientSubscriptions);
+        try {
+            return run(action, cancellationToken, listenerConfig, new PhasedActionEventConsumer(failsafePhasedActionResultListener, payloadSerializer, listenerConfig.buildEventConsumer),
+                providerParameters, params);
+        } catch (BuildExceptionVersion1 e) {
```

Comment:
⭕️ I think this logic should live in the phased build action runner (i.e. a try-catch around buildController.run())